Rewrite the .NET Aspire Azure OpenAI integration article (#2722)
dotnet/docs-aspire: docs/azureai/azureai-openai-integration.md

---
title: .NET Aspire Azure OpenAI integration (Preview)
description: Learn how to use the .NET Aspire Azure OpenAI integration.
ms.date: 03/04/2025
---

# .NET Aspire Azure OpenAI integration (Preview)

> Azure OpenAI Service provides access to OpenAI's powerful language models including GPT-4, GPT-3.5-Turbo, and Embeddings model series with the security and enterprise promise of Azure. The .NET Aspire Azure OpenAI integration enables you to connect to Azure OpenAI Service or OpenAI's API from your .NET applications.

Author: ```suggestion
[!INCLUDE [includes-hosting-and-client](../includes/includes-hosting-and-client.md)]

[Azure OpenAI Service](https://azure.microsoft.com/products/ai-services/openai-service) provides access to OpenAI's powerful language models including GPT-4, GPT-3.5-Turbo, and Embeddings model series with the security and enterprise promise of Azure. The .NET Aspire Azure OpenAI integration enables you to connect to Azure OpenAI Service or OpenAI's API from your .NET applications.
```

       .WithReference(openai);

// After adding all resources, run the app...
```

The preceding code adds an Azure OpenAI resource named `openai` to the app host project. The <xref:Aspire.Hosting.ResourceBuilderExtensions.WithReference%2A> method passes the connection information to the `ExampleProject` project.

> [!IMPORTANT]
> When you call <xref:Aspire.Hosting.AzureOpenAIExtensions.AddAzureOpenAI%2A>, it implicitly calls <xref:Aspire.Hosting.AzureProvisionerExtensions.AddAzureProvisioning(Aspire.Hosting.IDistributedApplicationBuilder)>—which adds support for generating Azure resources dynamically during app startup. The app must configure the appropriate subscription and location. For more information, see [Local provisioning: Configuration](../azure/local-provisioning.md#configuration).

### Add an Azure OpenAI deployment resource

To add an Azure OpenAI deployment resource, call the <xref:Aspire.Hosting.AzureOpenAIExtensions.AddDeployment(Aspire.Hosting.ApplicationModel.IResourceBuilder{Aspire.Hosting.ApplicationModel.AzureOpenAIResource},Aspire.Hosting.ApplicationModel.AzureOpenAIDeployment)> method:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

> var openai = builder.AddAzureOpenAI("openai").AddDeployment(
>     new AzureOpenAIDeployment(
>         name: "preview",
>         modelName: "gpt-4.5-preview",
>         modelVersion: "2025-02-27"));

Reviewer: We should split these lines up. Just like EventHubs, CosmosDB, ServiceBus, once [Configure OpenAI models in the app host (dotnet/aspire#6577)](https://github.com/dotnet/aspire/pull/6577) is completed the `AddDeployment` is going to return a "child resource" that points directly to the deployment "preview".

```suggestion
var openai = builder.AddAzureOpenAI("openai");
openai.AddDeployment(
    new AzureOpenAIDeployment(
        name: "preview",
        modelName: "gpt-4.5-preview",
        modelVersion: "2025-02-27"));
```

builder.AddProject<Projects.ExampleProject>()
       .WithReference(openai);

// After adding all resources, run the app...
```

The preceding code:

- Adds an Azure OpenAI resource named `openai`.
- Adds an Azure OpenAI deployment resource named `preview` with a model name of `gpt-4.5-preview`. The model name must correspond to an [available model](/azure/ai-services/openai/concepts/models) in the Azure OpenAI service.

### Generated provisioning Bicep

If you're new to [Bicep](/azure/azure-resource-manager/bicep/overview), it's a domain-specific language for defining Azure resources. With .NET Aspire, you don't need to write Bicep by-hand, instead the provisioning APIs generate Bicep for you. When you publish your app, the generated Bicep provisions an Azure OpenAI resource with standard defaults.

> :::code language="bicep" source="../snippets/azure/AppHost/openai.module.bicep":::

Reviewer: This bicep doesn't seem to align with the above call to `AddDeployment`. I don't see the model in it.

Author: That's a good catch. I'll fix that.

Author: Fixed

:::code language="csharp" source="../snippets/azure/AppHost/Program.ConfigureOpenAIInfra.cs" id="configure":::

The preceding code:

- Chains a call to the <xref:Aspire.Hosting.AzureProvisioningResourceExtensions.ConfigureInfrastructure*> API:
  - The `infra` parameter is an instance of the <xref:Aspire.Hosting.Azure.AzureResourceInfrastructure> type.
  - The provisionable resources are retrieved by calling the <xref:Azure.Provisioning.Infrastructure.GetProvisionableResources> method.
  - The single <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount> resource is retrieved.
  - The <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount.Sku?displayProperty=nameWithType> property is assigned to a new instance of <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSku> with an `E0` name and <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSkuTier.Enterprise?displayProperty=nameWithType> tier.
  - A tag is added to the Cognitive Services resource with a key of `ExampleKey` and a value of `Example value`.

> ### Connect to an existing Azure OpenAI service

Reviewer: Should this be updated to use the "AsExisting" functionality?

Author: I could update it, but there's a link to it after the code snippet - encouraging readers to learn more about those specific APIs.

Reviewer: AddConnectionString is weakly typed - it is just a string - you don't know what it is pointing to: a storage account or a CosmosDB?

AsExisting is typed, so you know that it is Storage.

When using Azure Functions, you'd want AsExisting because when you `WithReference` it will set the right configuration settings for Functions.

My thinking is that we should guide users to use AsExisting where possible, it is the better API.

Author: Fixed

You might have an existing Azure OpenAI service that you want to connect to. Instead of representing a new Azure OpenAI resource, you can add a connection string to the app host. To add a connection to an existing Azure OpenAI service, call the <xref:Aspire.Hosting.ParameterResourceBuilderExtensions.AddConnectionString%2A> method:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var openai = builder.ExecutionContext.IsPublishMode
    ? builder.AddAzureOpenAI("openai")
    : builder.AddConnectionString("openai");

builder.AddProject<Projects.ExampleProject>()
       .WithReference(openai);

// After adding all resources, run the app...
```

[!INCLUDE [connection-strings-alert](../includes/connection-strings-alert.md)]

The connection string is configured in the app host's configuration, typically under User Secrets, under the `ConnectionStrings` section:

```json
{
  "ConnectionStrings": {
    "openai": "https://{account_name}.openai.azure.com/"
  }
}
```

For more information on treating Azure OpenAI resources as existing resources, see [Use existing Azure resources](../azure/integrations-overview.md#use-existing-azure-resources).

> ### Hosting integration health checks
> 
> The Azure OpenAI hosting integration automatically adds a health check for the OpenAI resource. The health check verifies that the Azure OpenAI service is available and accessible.
> 

Reviewer: ```suggestion
```

This isn't true. There is no health check

- [Quickstart: Get started using GPT-35-Turbo and GPT-4 with Azure OpenAI Service](/azure/ai-services/openai/chatgpt-quickstart?pivots=programming-language-csharp).

### Add Azure OpenAI client with registered `IChatClient`

If you're interested in using the new <xref:Microsoft.Extensions.AI.IChatClient> interface, with the OpenAI client, simply chain either of the following APIs to the `AddAzureOpenAIClient` method:

- <xref:Microsoft.Extensions.Hosting.AspireOpenAIClientBuilderChatClientExtensions.AddChatClient(Aspire.OpenAI.AspireOpenAIClientBuilder,System.String)>: Registers a singleton `IChatClient` in the services provided by the <xref:Aspire.OpenAI.AspireOpenAIClientBuilder>.
- <xref:Microsoft.Extensions.Hosting.AspireOpenAIClientBuilderChatClientExtensions.AddKeyedChatClient(Aspire.OpenAI.AspireOpenAIClientBuilder,System.String,System.String)>: Registers a keyed singleton `IChatClient` in the services provided by the <xref:Aspire.OpenAI.AspireOpenAIClientBuilder>.

For example, consider the following C# code that adds an `IChatClient` to the DI container:

```csharp
builder.AddAzureOpenAIClient(connectionName: "openai")

>        .AddChatClient();

Reviewer: In 9.1 you need to pass a model name. That will be fixed once we have [Configure OpenAI models in the app host (dotnet/aspire#6577)](https://github.com/dotnet/aspire/pull/6577).

```suggestion
       .AddChatClient("deploymentName");
```

Author: According to our [API docs, it's optional](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.aspireopenaiclientbuilderchatclientextensions.addchatclient?view=dotnet-aspire-9.1#microsoft-extensions-hosting-aspireopenaiclientbuilderchatclientextensions-addchatclient(aspire-openai-aspireopenaiclientbuilder-system-string)).

Reviewer: See https://github.com/dotnet/aspire/blob/2a8f48ea5811f317a26405eb315aa315cc9e3cea/src/Components/Aspire.OpenAI/AspireOpenAIClientBuilderChatClientExtensions.cs#L20

>     /// <param name="deploymentName">Optionally specifies which model deployment to use. If not specified, a value will be taken from the connection string.</param>

```

Similarly, you can add a keyed `IChatClient` with the following C# code:

```csharp
builder.AddAzureOpenAIClient(connectionName: "openai")

>        .AddKeyedChatClient();

Reviewer: You need to at least pass a serviceKey

```suggestion
       .AddKeyedChatClient("serviceKey", "deploymentName");
```

---
title: .NET Aspire Azure OpenAI integration (Preview)
description: Learn how to use the .NET Aspire Azure OpenAI integration.
ms.date: 03/06/2025
---

# .NET Aspire Azure OpenAI integration (Preview)

[!INCLUDE [includes-hosting-and-client](../includes/includes-hosting-and-client.md)]

> [Azure OpenAI Service](https://azure.microsoft.com/products/ai-services/openai-service) provides access to OpenAI's powerful language models including GPT-4, GPT-3.5-Turbo, and Embeddings model series with the security and enterprise promise of Azure. The .NET Aspire Azure OpenAI integration enables you to connect to Azure OpenAI Service or OpenAI's API from your .NET applications.

Reviewer 2: I'd recommend not citing specific model names unless you're in a position to update this every month. What's written here already looks pretty outdated (3.5 Turbo, absence of mention of any o-series models).

> [Azure OpenAI Service](https://azure.microsoft.com/products/ai-services/openai-service) provides access to OpenAI's powerful language models including GPT-4, GPT-3.5-Turbo, and Embeddings model series with the security and enterprise promise of Azure. The .NET Aspire Azure OpenAI integration enables you to connect to Azure OpenAI Service or OpenAI's API from your .NET applications.

Author: ```suggestion
[Azure OpenAI Service](https://azure.microsoft.com/products/ai-services/openai-service) provides access to OpenAI's powerful language and embedding models with the security and enterprise promise of Azure. The .NET Aspire Azure OpenAI integration enables you to connect to Azure OpenAI Service or OpenAI's API from your .NET applications.
```

  - The `infra` parameter is an instance of the <xref:Aspire.Hosting.Azure.AzureResourceInfrastructure> type.
  - The provisionable resources are retrieved by calling the <xref:Azure.Provisioning.Infrastructure.GetProvisionableResources> method.
  - The single <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount> resource is retrieved.
  - The <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount.Sku?displayProperty=nameWithType> property is assigned to a new instance of <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSku> with an `E0` name and <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSkuTier.Enterprise?displayProperty=nameWithType> tier.
  - A tag is added to the Cognitive Services resource with a key of `ExampleKey` and a value of `Example value`.

### Connect to an existing Azure OpenAI service

You might have an existing Azure OpenAI service that you want to connect to. You can chain a call to annotate that your <xref:Aspire.Hosting.ApplicationModel.AzureOpenAIResource> is an existing resource:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var openai = builder.AddAzureOpenAI("openai")

>                     .AsExisting();

Reviewer: You need to pass 2 parameters into AsExisting to get the name and resource group.

> var openai = builder.AddAzureOpenAI("openai")
>                     .AsExisting();

Author: ```suggestion
var existingOpenAIName = builder.AddParameter("existingOpenAIName");
var existingOpenAIResourceGroup = builder.AddParameter("existingOpenAIResourceGroup");

var openai = builder.AddAzureOpenAI("openai")
                    .AsExisting(existingOpenAIName, existingOpenAIResourceGroup);
```

```csharp
builder.AddAzureOpenAIClient(connectionName: "openai");
```

> [!TIP]
> The `connectionName` parameter must match the name used when adding the Azure OpenAI resource in the app host project. For more information, see [Add an Azure OpenAI resource](#add-an-azure-openai-resource).

After adding the `OpenAIClient`, you can retrieve the client instance using dependency injection:

```csharp
public class ExampleService(OpenAIClient client)
{
    // Use client...
}
```

For more information, see:

- [Azure.AI.OpenAI documentation](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/openai/Azure.AI.OpenAI/README.md) for examples on using the `OpenAIClient`.
- [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection) for details on dependency injection.
- [Quickstart: Get started using GPT-35-Turbo and GPT-4 with Azure OpenAI Service](/azure/ai-services/openai/chatgpt-quickstart?pivots=programming-language-csharp).

### Add Azure OpenAI client with registered `IChatClient`

> If you're interested in using the new <xref:Microsoft.Extensions.AI.IChatClient> interface, with the OpenAI client, simply chain either of the following APIs to the `AddAzureOpenAIClient` method:

Reviewer 3: I don't like saying things like "The **new** blah interface." Eventually this won't be "new" which means you have to edit your doc to remove this type of wording.

> If you're interested in using the new <xref:Microsoft.Extensions.AI.IChatClient> interface, with the OpenAI client, simply chain either of the following APIs to the `AddAzureOpenAIClient` method:

Author: ```suggestion
If you're interested in using the <xref:Microsoft.Extensions.AI.IChatClient> interface, with the OpenAI client, simply chain either of the following APIs to the `AddAzureOpenAIClient` method:
```

title: .NET Aspire Azure OpenAI integration (Preview)
description: Learn how to use the .NET Aspire Azure OpenAI integration.

> ms.topic: how-to

Author: ```suggestion
```

- `Azure.Core`
- `Azure.Identity`

### Tracing

> The .NET Aspire Azure OpenAI integration emits tracing activities using OpenTelemetry for operations performed with the `OpenAIClient`.

Reviewer: The Azure tracing is still experimental, so it needs to be opted in.

https://github.com/openai/openai-dotnet/blob/3d7d89dce2a9b4c27708028f461df2f549ff11ad/docs/observability.md?plain=1#L13-L25

Also, we enable Microsoft.Extensions.AI tracing as well.

Thinking more about this, I wonder if we really want both, or if we should have separate switches for them. Thoughts @stephentoub @SteveSandersonMS? In the future, would someone want/need both sets turned on? Do we have guidance?

Reviewer 4: You probably want one or the other, but not both. The experimental tracing in OpenAI implements the otel standard convention for genai (or, at least, an older version of it), just as UseOpenTelemetry does.

Reviewer: Thanks! I opened [Tracing for OpenAI integrations (dotnet/aspire#7931)](https://github.com/dotnet/aspire/issues/7931) to decide and execute on out what we want to do here.

> The .NET Aspire Azure OpenAI integration emits tracing activities using OpenTelemetry for operations performed with the `OpenAIClient`.

Author: ```suggestion
The .NET Aspire Azure OpenAI integration emits tracing activities using OpenTelemetry for operations performed with the `OpenAIClient`. Tracing is currently experimental with this integration. To opt-in to it, set either the `OPENAI_EXPERIMENTAL_ENABLE_OPEN_TELEMETRY` environment variable to `true` or `1`, or by call `AppContext.SetSwitch("OpenAI.Experimental.EnableOpenTelemetry", true))` during app startup.
```
